Updates to allow installation on heroku

josegonzalez opened the pull request:
Note that this depends upon heroku/heroku-buildpack-python#271 to make cabot _fully_ heroku-compatible with the auth ldap backend.

TODO:
- [x] Fix tests
- [x] Add a django manage script to create the user based on the provided env vars
- [x] Cache nodejs/lessc requirement

josegonzalez commented:
Certainly possible I broke tests with my setup.py changes. Will run this locally and figure it out. I also currently consider this a work in progress, though welcome any and all feedback.

josegonzalez commented:
I only have one issue left here, but what are your thoughts on the changes, @dbuxton ?

JeanFred commented on bin/post_compile:
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env bash
```
This script only makes sense for heroku, correct? If so, does it have to be called `post_compile` or can we make it clear it is  for Heroku only ?

josegonzalez replied:
Only heroku seems to ever call `bin/post_compile` and I've added a check for the `DYNO` env var that it sets

josegonzalez commented on bin/post_compile:
```diff
@@ -0,0 +1,99 @@
+
+
+main() {
+  if [[ ! -n $DYNO ]]; then
```
This check here ensures none of this is run outside of heroku. We can probably add some sort of docblock to the top of this file to signify heroku running only though.

josegonzalez commented on cabot/urls.py:
```diff
@@ -140,6 +140,9 @@
      url(r'^admin/', include(admin.site.urls)),
 
+     # for serving static assets using django
+     url(r'^static/(?P<path>.*)$', 'django.views.static.serve', {'document_root': settings.STATIC_ROOT,}),
```
In theory if you are using a webserver like nginx, you can skip this completely for non-heroku deploys. Should that be documented if it already isn't? I think no, but up to you.

moubry commented:
I can’t wait until this is merged. I’d love to run Cabot on Heroku!

josegonzalez commented:
I have to rebase for the recent `requests` change, and need a bit of help figuring out the custom django command (I've never used django before).

dbuxton commented:
@josegonzalez the management commands are pretty easy as long as you don't need to pass too many options (you can see that they can get pretty complicated though by looking at [`createsuperuser`](https://github.com/django/django/blob/master/django/contrib/auth/management/commands/createsuperuser.py)!)

You need to do something like:

```
from django.contrib.auth.models import User
from django.core.management.base import BaseCommand
import os

class Command(BaseCommand):
    help = 'Automatically create superuser based on environment variables.'
    def handle(self, *args, **options):
        username = os.getenv('CABOT_ADMIN_EMAIL')
        u = User(username=username, email=username)
        u.set_password(os.getenv('CABOT_ADMIN_PASSWORD'))
        u.is_superuser = True
        u.is_staff = True
        u.save()
```

Hope that helps

josegonzalez commented:
@dbuxton where do I put that in the app path?

dbuxton commented:
`cabot/cabotapp/management/commands/create_superuser_on_heroku.py`, obviously ;)

Then it will be callable as `./manage.py create_superuser_on_heroku`.

josegonzalez commented:
Mailed it.

kevinprince commented:
Any plans to merge this?

dbuxton commented:
It's a pretty big changeset - I want to verify that deployment works out of the box for bare metal/cloud install first. (At least for digitalocean)

josegonzalez commented:
@dbuxton Any update on that?

josegonzalez commented:
@dbuxton I tested this against a new digital ocean instance, seems to work just fine. Also rebased against master and cleaned up a bit of the bash code.

Would love it if you could merge this/let me know whats left before it can be merged :)

josegonzalez commented:
*bump*

dbuxton commented:
@JeanFred can you have a look?

josegonzalez commented:
Just rebased this again :)

JeanFred commented:
Sorry for the delay in getting back to you @josegonzalez . I will aim to review this in the coming week. Thanks for your patience!

josegonzalez commented:
I'll rebase this again to pre-empt the inevitable "please rebase" response :)

JeanFred commented on cabot/cabotapp/management/commands/create_superuser_on_heroku.py:
```diff
@@ -0,0 +1,21 @@
+from django.contrib.auth.models import User
```
This management command sounds useful with or without heroku. Can we rename it `create_superuser_from_environment_variables.py` ?

josegonzalez replied:
Done.

josegonzalez commented:
Rebased.

JeanFred commented:
Thanks!

I tried to pull that locally and it fails to compile (for the same reasons that the Travis is failing). I cannot import the ldap_settings, because these import the ldap package which fails.

On master, `django-auth-ldap` is included in `requirements.txt`, which is not the case here (it is conditionally included in setup.py depending on an env var. This would not play well with a pure `pip install requirements.txt`

Not sure how elegant this was in the first place (installing by default a package only useful for some) but for now I’ll just move it back into `requirements.txt`.

josegonzalez commented:
@JeanFred The ldap pr wasn't merged into heroku, so installing with ldap won't work.

josegonzalez commented:
I'll look at the test fails. I basically want ldap to be optional as heroku doesn't support it - I wasn't able to continue my pr.

frankh commented:
Cleaning up old PRs

Feel free to re-open with the conflicts resolved if you think it's still relevant
